File: backend/cxx/src/sp.cxx
bool SPFamilies::containsFamily(FamilyID id) const { return families.count(id); }
Node * SPFamilies::getFamily(FamilyID id) const
{
  assert(families.count(id));
  return families[id];
}

void SPFamilies::registerFamily(FamilyID id,RootNodePtr root)
{
  assert(!families.count(id));
  families[id] = root;
}

void SPFamilies::unregisterFamily(FamilyID id)
{
  assert(families.count(id));
  families.erase(id);
}

shared_ptr<SPAux> SPAux::copy() { return new SPAux(); }

shared_ptr<SPAux> VentureSP::constructSPAux() const { return new SPAux(); }
shared_ptr<LatentDB> VentureSP::constructLatentDB() const { return NULL; }

void VentureSP::simulateLatents(SPAux * spaux,LSR * lsr,bool shouldRestore,LatentDB * latentDB) const { throw "no default latent handling"; }
double VentureSP::detachLatents(SPAux * spaux,LSR * lsr,LatentDB * latentDB) const { throw "no default latent handling"; }
bool VentureSP::hasAEKernel() const { return false; }

File: backend/jventure/sp.jl
require("psp.jl")

type SP
  requestPSP::RequestPSP
  outputPSP::OutputPSP
  name::String
end

type SPRef
  makerNode::Union(ConstantNode,OutputNode)
end

type SPRecord
  sp::Union(SP,Nothing)
  aux::Any
  families::Dict{SPFamilyID,Node}
end



  # def constructLatentDB(self): return None
  # def simulateLatents(self,spaux,lsr,shouldRestore,latentDB): pass
  # def detachLatents(self,spaux,lsr,latentDB): pass
  # def hasAEKernel(self): return False